src/util/stream.rs

```diff
@@ -38,3 +44,37 @@ macro_rules! ready {
         }
     };
 }
+
+pub async fn tcp_connect_with_timeout<T>(addr: T, request_timeout_s: u64) -> Result<TcpStream>
+    where T: ToSocketAddrs,
+{
+    let fut = tcp_connect(addr);
+    match timeout(Duration::from_secs(request_timeout_s), fut).await {
+        Ok(result) => result,
+        Err(_) => Err(ReplyError::TtlExpired.into()),
```

Comment:
I wonder if we should create a new kind of error, eg. `ConnectTimeout`? What do you think?

Reply:
Makes sense to me. Going to add it.

Reply:
Added it. There doesn't seem to be a specific Socks5 error we can make to (going from the spec), so I still map it to TTL expired.